File: src/main/kotlin/xyz/mahmoudahmed/dsl/TypedBranchMappingBuilder.java
package xyz.mahmoudahmed.dsl;

import java.util.function.BiConsumer;
import java.util.function.Function;

public class TypedBranchMappingBuilder<S, T, V> {
    private final TypedBranchActionBuilder<S, T> parent;
    private final Function<S, V> sourceAccessor;
    private Function<V, Object> transformation;

    TypedBranchMappingBuilder(TypedBranchActionBuilder<S, T> parent, Function<S, V> sourceAccessor) {
        this.parent = parent;
        this.sourceAccessor = sourceAccessor;
    }

    public <U> TypedBranchTargetBuilder<S, T, V, U> to(BiConsumer<T, U> targetSetter) {
        return new TypedBranchTargetBuilder<>(parent, sourceAccessor, targetSetter, transformation);
    }

    public TypedBranchMappingBuilder<S, T, V> transform(Function<V, Object> transformation) {
        this.transformation = transformation;
        return this;
    }
}

File: src/main/kotlin/xyz/mahmoudahmed/dsl/typed/TypedBranchMappingBuilder.kt
package xyz.mahmoudahmed.dsl.typed


class TypedBranchMappingBuilder<S, T, V>(
    private val parent: TypedBranchActionBuilder<S, T>,
    private val sourceAccessor: (S) -> V
) {
    private var transformation: ((V) -> Any)? = null

    fun <U> to(targetSetter: (T, U) -> Unit): TypedBranchTargetBuilder<S, T, V, U> {
        return TypedBranchTargetBuilder(parent, sourceAccessor, targetSetter, transformation)
    }

    fun transform(transformation: (V) -> Any): TypedBranchMappingBuilder<S, T, V> {
        this.transformation = transformation
        return this
    }
}